### PR TITLE
Logarithmic encoding and normalization

### LINKED ISSUE #99: Fit logarithmic data
## Description
Currently `fit` fails with logarithmic input data and logarithmic outputs.

## Example

log_fit_210423

log_fit2_210423

## Workaround
Modify input and output data manually / with postprocessor

### LINKED ISSUE #102: Fits are not saved in default settings
save=model.hdf5 seems not to be the default in fits, but rather `False` (?). Confused me when running profit ui on mockup and it could not find the saved fit.

### LINKED ISSUE #106: Fit does not work for small length scales
I can't get the `fit` to work for the new data generated with `GORILLA`.

**Version:** the tests were conducted using the latest (as of 2021-04-24) version of `mkendler` as seen in PR #97. In previous versions the fit also sometimes produced a similar result.

This bug might be related to #99. It might even be that this bug is actually responsible for the behavior seen in #99 and not the logarithmic scale.

# output of `fit` - convergence seems to be a problem

<details>
<summary> no configuration </summary>

```
Initial hyperparameters:
length_scale: [0.00151388]
sigma_n: [0.00024184]
sigma_f: [0.00634512]
Optimized hyperparameters:
length_scale: [21.10524403]
sigma_n: [0.01127425]
sigma_f: [0.08146067]
```

</details>
<details>
<summary> configuring 2 length scales </summary>

```
Initial hyperparameters:
length_scale: [0.001  0.0001]
sigma_n: [0.00024184]
sigma_f: [0.00634512]
Optimized hyperparameters:
length_scale: [34.29999082 28.57668733]
sigma_n: [0.01127386]
sigma_f: [0.0814779]
```
</details>

# Reduction to one dimension
No inital `length_scale` specified, just one input variable, one output variable

<details>
<summary> fit output looks better </summary>

```
Initial hyperparameters:
length_scale: [0.00143508]
sigma_n: [5.18076353e-08]
sigma_f: [1.38748647e-06]
Optimized hyperparameters:
length_scale: [0.00182629]
sigma_n: [3.97746379e-08]
sigma_f: [7.6791768e-08]
```

</details>
<details>
<summary> result </summary>

image

</details>

# Data

<details>
<summary> reduced to 1D </summary>

**input**
```
# nu_star
5.500000000000000548e-03
3.250000000000000284e-03
7.750000000000000812e-03
2.125000000000000153e-03
6.625000000000000680e-03
4.375000000000000416e-03
8.875000000000000944e-03
1.562500000000000087e-03
6.062500000000001048e-03
3.812500000000000350e-03
8.312500000000000444e-03
2.687500000000000219e-03
7.187500000000000312e-03
4.937500000000000049e-03
9.437500000000001443e-03
1.281250000000000054e-03
5.781250000000000798e-03
3.531250000000000534e-03
8.031250000000000194e-03
2.406250000000000402e-03
```

**output**
```
# D11
7.232595332146684520e-06
7.626592467180592352e-06
1.008375721371916446e-05
9.751873982459288823e-06
6.427973623514244548e-06
6.727557691294112411e-06
6.422614234207888881e-06
6.253991403170845712e-06
6.546003910961603981e-06
6.959192715269715775e-06
7.324520607679214362e-06
7.952616716490858097e-06
7.800422060347728914e-06
7.480339311007465075e-06
1.143475493108907343e-05
9.034020210949410081e-06
7.072863504479973163e-06
7.024873569713297894e-06
9.595086251789200965e-06
8.159805847859729049e-06
```

</details>

# Rescaling the variables
I tried rescaling both variables with `new = old / old.mean()` and this actually gives meaningful results (in the 1D case)!

<details>
<summary> fit output </summary>

```
Initial hyperparameters:
length_scale: [0.27722608]
sigma_n: [0.00660342]
sigma_f: [0.17684961]
Optimized hyperparameters:
length_scale: [1.52715434]
sigma_n: [0.1635188]
sigma_f: [1.22208624]
```

</details>
<details>
<summary> result </summary>

image

</details>

using the same approach in 2D works as well, although the fit is *too good* in this case.

<details>
<summary> result 2D </summary>

image

</details>

### PR DESCRIPTION
**Pull request by mkendler**
* Encoder class
  * [x] Normalization of input and output data (Fix #106)
  * [x] Log transform `LogUniform` input variables (Close #99)
* Smaller fixes
  * [x] Save model by default (Fix #102)
  * [x] Remove Tensorflow references
  * [x] Smaller bug fixes
* [x] Add list of available classes (Surrogates, runners, interfaces, etc.) in docs
* [x] Unit tests for surrogate functions

### CODE REVIEW

**Comment by mkendler** [comment]
@Rykath, I added the default runners, interfaces, etc. in the Config doc. Please check if their description is correct.

File: profit/sur/gaussian_process.py [hunk 1]
```diff
@@ -482,6 +480,7 @@ def optimize(self, fixed_sigma_n=False, eval_gradient=False, return_hess_inv=Fal
         self.hyperparameters['sigma_f'] = np.atleast_1d(opt_hyperparameters[last_idx])
         if not self.fixed_sigma_n:
             self.hyperparameters['sigma_n'] = np.atleast_1d(opt_hyperparameters[-1])
+        self.print_hyperparameters("Optimized")
```
Comment (krystophny) [inline profit/sur/gaussian_process.py:483]: In the future, such input/output should not be done inside the fitting code to decouple interface from core.

Comment (krystophny) [reply profit/sur/gaussian_process.py:483]: ... one way to do this would be a proper event architecture, where the console (or web) UI can pass an event handler that receives a message when hyperparameters are optimized, together with their values. The display logic is then handled by the respective UI frontend.

File: profit/sur/sur.py [hunk 2]
```diff
@@ -29,31 +29,97 @@ class Surrogate(ABC):
...
+
+        Parameters:
+            x (ndarray): Prediction input points.
+
+        Returns:
+            ndarray: Encoded and normalized prediction points.
+        """
+        for enc in self.encoder:
+            if not enc.output:
+                if enc.label == 'normalization':
```
Comment (krystophny) [inline profit/sur/sur.py:95]: I guess this will be changed to nice decoupled object-oriented design later, right?

**Review by krystophny — APPROVED** [review]

**Comment by Rykath** [comment]
> @Rykath, I added the default runners, interfaces, etc. in the Config doc. Please check if their description is correct.

Well, `zeromq` is not quite right but overall it's ok, thanks. I will fix it in my own PR. However I have actually documented all available components farther down right where the respective option is documented. This also tells the users where the component needs to be selected. I do think that a description and status / drawback information would be useful.

**Comment by krystophny** [comment]
Oh. I didn't get that you were still discussing, but thanks for doing this in the other PR then!

**Comment by Rykath** [comment]
@mkendler, I also think the description of the variable kinds could be a bit expanded. Maybe add each variable kind to the example.

Also we might want to discuss whether it's not more useful to specify independent variables using `size` instead of `step`. This is also the usual parameter for `linspace`.

File: profit/sur/gaussian_process.py (hunk 1 above)
Comment (mkendler) [reply profit/sur/gaussian_process.py:483]: This is a very good idea. We should discuss if a general event-driven architecture should be one of our main focuses for v0.5.

File: profit/sur/sur.py (hunk 2 above)
Comment (mkendler) [reply profit/sur/sur.py:95]: Exactly, the idea is to set up a modular `Encoder` class just like the surrogates, where also custom encoder functions can be registered. This will be done in v0.5.
